pkg/controllers/deprovisioning/controller.go

```diff
@@ -105,126 +108,64 @@ func (c *Controller) Builder(_ context.Context, m manager.Manager) controller.Bu
-	*v1.Node
-	instanceType   *cloudprovider.InstanceType
-	capacityType   string
-	zone           string
-	provisioner    *v1alpha5.Provisioner
-	disruptionCost float64
-	pods           []*v1.Pod
-}
-
-// ProcessCluster is exposed for unit testing purposes
-// ProcessCluster loops through implemented deprovisioners
-func (c *Controller) ProcessCluster(ctx context.Context) (Result, error) {
-	// range over the different deprovisioning methods. We'll only let one method perform an action
-	for _, d := range []Deprovisioner{
-		// Expire any nodes that must be deleted, allowing their pods to potentially land on currently
-		// empty nodes
-		c.expiration,
-
-		// Terminate any nodes that have drifted from provisioning specifications, allowing the pods to reschedule.
-		c.drift,
-
-		// Delete any remaining empty nodes as there is zero cost in terms of dirsuption.  Emptiness and
-		// emptyNodeConsolidation are mutually exclusive, only one of these will operate
-		c.emptiness,
-		c.emptyNodeConsolidation,
-
-		// Attempt to identify multiple nodes that we can consolidate simultaneously to reduce pod churn
-		c.multiNodeConsolidation,
-
-		// And finally fall back our single node consolidation to further reduce cluster cost.
-		c.singleNodeConsolidation,
-	} {
+	// Attempt different deprovisioning methods. We'll only let one method perform an action
+	for _, d := range c.deprovisioners {
 		candidates, err := candidateNodes(ctx, c.cluster, c.kubeClient, c.clock, c.cloudProvider, d.ShouldDeprovision)
 		if err != nil {
-			return ResultFailed, fmt.Errorf("determining candidate nodes, %w", err)
+			return reconcile.Result{}, fmt.Errorf("determining candidate nodes, %w", err)
 		}
 		// If there are no candidate nodes, move to the next deprovisioner
 		if len(candidates) == 0 {
 			continue
 		}
 
-		result, err := c.executeDeprovisioning(ctx, d, candidates...)
+		// Determine the deprovisioning action
+		cmd, err := d.ComputeCommand(ctx, candidates...)
 		if err != nil {
-			return ResultFailed, fmt.Errorf("deprovisioning nodes, %w", err)
+			return reconcile.Result{}, fmt.Errorf("computing deprovisioning decision, %w", err)
 		}
-
-		switch result {
-		case ResultFailed:
-			return ResultFailed, err
-		case ResultRetry, ResultSuccess:
-			// the controller wants to retry, or was successful in deprovisioning
-			return result, nil
-		case ResultNothingToDo:
-			// found nothing to do, so try the next deprovisioner
+		if cmd.action == actionDoNothing {
 			continue
-		default:
-			logging.FromContext(ctx).Errorf("unexpected result %s", result)
 		}
+		if cmd.action == actionRetry {
+			return reconcile.Result{Requeue: true}, nil
+		}
+
+		// Attempt to deprovision
+		if err := c.executeCommand(ctx, d, cmd); err != nil {
+			return reconcile.Result{}, fmt.Errorf("deprovisioning nodes, %w", err)
+		}
+		return reconcile.Result{Requeue: true}, nil
 	}
 
+	// Mark cluster as consolidated if all deprovisioners succeeded
+	c.cluster.Consolidated()
```

Comment:
```suggestion
	c.cluster.SetConsolidated(true)
```

This returns whether the cluster is consolidated; I think you meant to mark that you have consolidated.